addons/account/models/account_move.py

```diff
@@ -1,349 +1,1896 @@
+    def _onchange_invoice_date(self):
+        if self.invoice_date:
+            self.invoice_date_due = self.date = self.invoice_date
+            self._onchange_currency()
+
+    @api.onchange('invoice_date_due')
+    def _onchange_invoice_date_due(self):
+        self._recompute_dynamic_lines()
+
+    @api.onchange('journal_id')
+    def _onchange_journal(self):
+        if self.journal_id:
+            new_currency = self.journal_id.currency_id or self.journal_id.company_id.currency_id
+            if new_currency != self.currency_id:
+                self.currency_id = new_currency
+                self._onchange_currency()
+
+    @api.onchange('partner_id')
+    def _onchange_partner_id(self):
+        for line in self.line_ids:
+            line.partner_id = self.commercial_partner_id
+
+        if self.type in ('out_invoice', 'out_refund', 'out_receipt'):
+            self.invoice_payment_term_id = self.partner_id.property_payment_term_id
+        elif self.type in ('in_invoice', 'in_refund', 'in_receipt'):
+            self.invoice_payment_term_id = self.partner_id.property_supplier_payment_term_id
+
+        # Find the new fiscal position.
+        delivery_partner_id = self._get_invoice_delivery_partner_id()
+        new_fiscal_position_id = self.env['account.fiscal.position'].get_fiscal_position(
+            self.partner_id.id, delivery_id=delivery_partner_id)
+        self.fiscal_position_id = self.env['account.fiscal.position'].browse(new_fiscal_position_id)
+        self._recompute_dynamic_lines()
+
+    @api.onchange('date', 'currency_id')
+    def _onchange_currency(self):
+        company_currency = self.company_id.currency_id
+        has_foreign_currency = self.currency_id and self.currency_id != company_currency
+
+        for line in self.line_ids:
+            new_currency = has_foreign_currency and self.currency_id
+            line.currency_id = new_currency
+            line._onchange_currency()
+        self._recompute_dynamic_lines()
+
+    @api.onchange('invoice_payment_ref')
+    def _onchange_invoice_payment_ref(self):
+        for line in self.line_ids.filtered(lambda line: line.account_id.user_type_id.type in ('receivable', 'payable')):
+            line.name = self.invoice_payment_ref
+
+    @api.onchange('invoice_vendor_bill_id')
+    def _onchange_invoice_vendor_bill(self):
+        if self.invoice_vendor_bill_id:
+            # Copy invoice lines.
+            for line in self.invoice_vendor_bill_id.invoice_line_ids:
+                copied_vals = line.copy_data()[0]
+                copied_vals['move_id'] = self.id
+                new_line = self.env['account.move.line'].new(copied_vals)
+                new_line.recompute_tax_line = True
+
+            # Copy payment terms.
+            self.invoice_payment_term_id = self.invoice_vendor_bill_id.invoice_payment_term_id
+
+            # Copy currency.
+            if self.currency_id != self.invoice_vendor_bill_id.currency_id:
```

Comment:
copy the currency before copying the lines, as it can trigger a bunch of useless onchanges

Reply:
Recompute is done with _recompute_dynamic_lines, a few lines after, so it would not change anything.